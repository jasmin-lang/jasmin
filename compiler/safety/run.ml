open Jasmin
open Jasmin_checksafety

let params =
  [
    ("success/loop2.jazz", "poly1305>in;");
    ("success/loop3.jazz", "poly1305>in;");
    ("fail/popcnt.jazz", "off_by_one>;");
  ]

module Arch =
  (val let use_set0 = !Glob_options.set0 and use_lea = !Glob_options.lea in
       let call_conv = !Glob_options.call_conv in
       let module C =
         (val CoreArchFactory.core_arch_x86 ~use_lea ~use_set0 call_conv)
       in
       (module Arch_full.Arch_from_Core_arch (C) : Arch_full.Arch
         with type reg = X86_decl.register
          and type regx = X86_decl.register_ext
          and type xreg = X86_decl.xmm_register
          and type rflag = X86_decl.rflag
          and type cond = X86_decl.condt
          and type asm_op = X86_instr_decl.x86_op
          and type extra_op = X86_extra.x86_extra_op))

let load_file name =
  let open Pretyping in
  name
  |> tt_file Arch.arch_info Env.empty None None
  |> fst |> Env.decls
  |> Compile.preprocess Arch.reg_size Arch.asmOp

let analyze ~fmt pd asmOp source_f_decl f_decl p =
  let module AbsInt = SafetyInterpreter.AbsAnalyzer (struct
    let main_source = source_f_decl
    let main = f_decl
    let prog = p
  end) in
  AbsInt.analyze ~fmt pd asmOp ()

let load_and_analyze ~fmt expect path name =
  let name = Filename.concat path name in
  Format.fprintf fmt "File %s:@." name;
  Glob_options.safety_param := List.assoc_opt name params;
  let ((_, fds) as p) = load_file name in
  List.iter
    (fun fd ->
      if FInfo.is_export fd.Prog.f_cc then
        let () =
          Format.fprintf fmt "@[<v>Analyzing function %s@]" fd.f_name.fn_name
        in
        let safe = analyze ~fmt Arch.pointer_data Arch.asmOp fd fd p in
        assert (safe = expect))
    fds

let doit ~fmt expect path =
  let cases = Sys.readdir path in
  Array.sort String.compare cases;
  Array.iter (load_and_analyze ~fmt expect path) cases

let () =
  let fmt = Format.std_formatter in
  doit ~fmt false "fail";
  (* Discard messages from successful tests *)
  let fmt = Format.make_formatter (fun _ _ _ -> ()) (fun () -> ()) in
  doit ~fmt true "success"
