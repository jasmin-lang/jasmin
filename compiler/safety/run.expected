File fail/arr_init_fail.jazz:
Analyzing function main


*** Possible Safety Violation(s):
  "fail/arr_init_fail.jazz", line 16 (4-15): in_bound: tmp[j] (length 16 U8)
  "fail/arr_init_fail.jazz", line 14 (2) to line 18 (3): termination
  
Memory ranges:
  mem_len: [0; 0]
  mem_x0: [0; 0]
  mem_x1: [0; 0]
  mem_x2: [0; 0]
  mem_x3: [0; 0]
  
* Rel:
{inv_len ≤ 18446744073709551615, inv_x3 ≥ 0, inv_x2 ≥ 0, inv_len ≥ 0, inv_x1 ≥ 0, inv_x0 ≥ 0, inv_x3 ≤ 4294967295, inv_x2 ≤ 4294967295, inv_x1 ≤ 4294967295, inv_x0 ≤ 4294967295, mem_x3 = 0, mem_x2 = 0, mem_x1 = 0, mem_x0 = 0, mem_len = 0}
mem_len ∊ [0; 0]
mem_x0 ∊ [0; 0]
mem_x1 ∊ [0; 0]
mem_x2 ∊ [0; 0]
mem_x3 ∊ [0; 0]



Program is not safe!

File fail/array-export.jazz:
Analyzing function main


*** Possible Safety Violation(s):
  "fail/array-export.jazz", line 4 (2-11): in_bound: a[i] (length 4 U8)
  
Memory ranges:
  mem_a: [0; 0]
  mem_i: [0; 0]
  
* Rel:
{inv_i ≤ 18446744073709551615, inv_i ≥ 0, mem_a = 0, mem_i = 0}
mem_a ∊ [0; 0]
mem_i ∊ [0; 0]



Program is not safe!

File fail/array-len3.jazz:
Analyzing function array


*** Possible Safety Violation(s):
  "fail/array-len3.jazz", line 33 (4-18): is_init blk[i]
  
* Rel:
⊤



Program is not safe!

File fail/array_init.jazz:
Analyzing function bad_read


*** Possible Safety Violation(s):
  "fail/array_init.jazz", line 8 (2-11): is_init t[0]
  
* Rel:
⊤



Program is not safe!

File fail/bad_align.jazz:
Analyzing function main


*** Possible Safety Violation(s):
  "fail/bad_align.jazz", line 6 (2-30): aligned pointer (pt +64u ((64u) 1)) u64
  
Memory ranges:
  mem_pt: [0; 9]
  
* Rel:
{mem_pt ≥ 0, inv_pt ≥ 0, mem_pt ≤ 9, inv_pt ≤ 18446744073709551615}
mem_pt ∊ [0; 9]

* Alignment: pt 64; 

Program is not safe!

File fail/bad_align2.jazz:
Analyzing function main


*** Possible Safety Violation(s):
  main return: is_init tmp
  
Memory ranges:
  mem_pt: [0; 120]
  
* Rel:
{mem_pt ≤ 120, inv_pt ≤ 18446744073709551615, inv_pt ≥ 0, mem_pt ≥ 0}
mem_pt ∊ [0; 120]



Program is not safe!

File fail/bounded_while.jazz:
Analyzing function uninit


*** Possible Safety Violation(s):
  "fail/bounded_while.jazz", line 15 (2-17): is_init t[(4 - 1)]
  
Memory ranges:
  mem_bound: [0; 0]
  
* Rel:
{inv_bound ≤ 18446744073709551615, inv_bound ≥ 0, mem_bound = 0}
mem_bound ∊ [0; 0]



Program is not safe!

File fail/carry_flags_err.jazz:
Analyzing function test


*** Possible Safety Violation(s):
  "fail/carry_flags_err.jazz", line 9 (4-27): termination
  
Bottom ⊥


Program is not safe!

File fail/carry_flags_err2.jazz:
Analyzing function test


*** Possible Safety Violation(s):
  "fail/carry_flags_err2.jazz", line 9 (4-27): termination
  
Bottom ⊥


Program is not safe!

File fail/decompile_fail.jazz:
Analyzing function test


*** Possible Safety Violation(s):
  "fail/decompile_fail.jazz", line 11 (4-11): is_init i
  "fail/decompile_fail.jazz", line 9 (2) to line 12 (3): termination
  
Bottom ⊥


Program is not safe!

File fail/fail_land.jazz:
Analyzing function BS2POLVECq


*** Possible Safety Violation(s):
  "fail/fail_land.jazz", line 7 (2-25): termination
  
Memory ranges:
  mem_b: [0; 0]
  
* Rel:
{inv_b ≤ 18446744073709551615, inv_b ≥ 0, mem_b = 0}
mem_b ∊ [0; 0]



Program is not safe!

File fail/fail_land2.jazz:
Analyzing function BS2POLVECq


*** Possible Safety Violation(s):
  "fail/fail_land2.jazz", line 6 (2-25): termination
  
Memory ranges:
  mem_b: [0; 0]
  
* Rel:
{inv_b ≥ 0, inv_b ≤ 18446744073709551615, mem_b = 0}
mem_b ∊ [0; 0]



Program is not safe!

File fail/init-join.jazz:
Analyzing function main


*** Possible Safety Violation(s):
  main return: is_init r
  
Memory ranges:
  mem_x: [0; 0]
  
* Rel:
{inv_x ≥ 0, inv_x ≤ 18446744073709551615, mem_x = 0}
mem_x ∊ [0; 0]



Program is not safe!

File fail/load.jazz:
Analyzing function load_last


*** Possible Safety Violation(s):
  "fail/load.jazz", line 15 (4-17): in_bound: s[:u8 j] (length 16 U8)
  "fail/load.jazz", line 19 (2-17): in_bound: s[:u8 j] (length 16 U8)
  
Memory ranges:
  mem_input: [0; 18446744073709551615]
  mem_len: [0; 0]
  
* Rel:
{inv_len ≤ 18446744073709551615, inv_input ≤ 18446744073709551615, inv_input ≥ 0, inv_len ≥ mem_input, mem_input ≥ 0, mem_len = 0}
mem_input ∊ [0; 18446744073709551615]
mem_len ∊ [0; 0]



Program is not safe!

File fail/loop2.jazz:
Analyzing function poly1305


*** Possible Safety Violation(s):
  "fail/loop2.jazz", line 8 (2) to line 20 (3): termination
  
Memory ranges:
  mem_in: [0; 18446744073709551615]
  mem_inlen: [0; 0]
  
* Rel:
{inv_inlen ≥ mem_in, inv_inlen ≤ 18446744073709551615, inv_in ≤ 18446744073709551615, inv_in ≥ 0, mem_in ≥ 0, mem_inlen = 0}
mem_in ∊ [0; 18446744073709551615]
mem_inlen ∊ [0; 0]



Program is not safe!

File fail/mul0.jazz:
Analyzing function addNumber


*** Possible Safety Violation(s):
  "fail/mul0.jazz", line 12 (5-28): termination
  
Bottom ⊥


Program is not safe!

File fail/shld.jazz:
Analyzing function undefined


*** Possible Safety Violation(s):
  "fail/shld.jazz", line 2 (2-30): (((uint) ((8u) 17)) % 32) ∈ [0; 16]
  
Memory ranges:
  mem_x: [0; 0]
  mem_y: [0; 0]
  
* Rel:
{inv_y ≥ 0, inv_x ≥ 0, inv_x ≤ 65535, inv_y ≤ 65535, mem_x = 0, mem_y = 0}
mem_x ∊ [0; 0]
mem_y ∊ [0; 0]



Program is not safe!

File fail/wint_overflow.jazz:
Analyzing function wint_overflow


*** Possible Safety Violation(s):
  "fail/wint_overflow.jazz", line 3 (28-48): 137 ∈ [-128; 127]
  
* Rel:
⊤



Program is not safe!

